### PR TITLE
Analytical sedov

### PR DESCRIPTION
**Pull request by jaescartin1**
Added analytical solution for the Sedov test.

It generate a file with the solution in 1D for the next geometries: planar(1D), cylindrical(2D) or spherical(3D)

### CODE REVIEW

**Comment by sekelle** [comment]
Hi Jose, thanks for your work. I'll be excited to see the comparison with an analytical reference.
I think the analytical solution should be an independent app, not coupled to `sedov.cpp` . You could then have a python or bash script that combines the data from the sedov app and the analytical reference and compare them.

**Comment by jaescartin1** [comment]
Hi Sebastian, the comparative is not completely ready yet. The solution works with a planar (1D), cylindrical (2D) or spherical (3D) symmetry.  It means that we still need to modify the sedov sph-exa input to provide a spherical symmetry 3D model. This is the previous step to work with Ruben an reach the final solution.

**Comment by jaescartin1** [comment]
I attached the original fortran code , fully tested in the folder 'src/sedov/analytical', we should remove this folder in the future.
The 'SedovAnalyticalSolution.hpp' file is my c++ adaptation from the original code.

**Comment by jaescartin1** [comment]
Ruben asked me to integrate the Sedov solution inside the sph-exa code, in order to generate automatically the solution when the user select dump and provide the solution flag. It is much more easy and useful to compare the simulate and the analytical output because the solution use automatically the same input parameters than sph-exa on that iteration and time.

**Comment by jaescartin1** [comment]
If you think that everything seems correct, and you don't mind, please approve the pull request and we can modify the implementation in the future if it is need. I would like to have it in the develop branch for the moment to test it the code with Ruben.

**Comment by jaescartin1** [comment]
Hi Sebastian, the Sedov exact solution is in place and working. We can merge the branch.

**Comment by jgphpc** [comment]
cscs-ci run

**Comment by jgphpc** [comment]
Some tests failed the CI: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/SPH-EXA-44692846847247/-/pipelines/452889762

**Comment by jaescartin1** [comment]
Hi Jean-Guillaume,
      thanks for check it.  I just removed a function to reduce the size to a sphere. It seems didn't match well with the GPU implementation after the reduccion. Like this function is not needed for the test, I just removed it.

I hope now the modifications pass the tests.

**Comment by jgphpc** [comment]
cscs-ci run

**Comment by sebkelle1** [comment]
Hi @jaescartin1, thanks for your contribution.
The analytical solution itself and the changes to the SedovDataGenerator are fine.

One problem that remains is that the code for the analytical solution is introduced into the main compilation unit of sedov.cpp.
We have to avoid this coupling, because what's now just the sedov test case in sedov.cpp will evolve into a more general simulation driver that will be able to handle any hydro-only type of simulation, including the turbulence tests on LUMI.
We don't actually want to repeat the whole time-stepping loop for every new set of initial conditions. Therefore we can't have
code for the analytical solution introduced in sedov.cpp

The sedov analytical solution can and should be a stand-alone application: You first execute a run with the sedov application, writing snapshots at the times where you want to compare with the analytical solutions.
Next, `sedov-analytical` can be executed. It includes the same sedov_data_generator.hpp file as sedov.cpp to get the simulation parameters and it reads the required simulation data (x,y,z,vx,vy,vz,time) from the output that the sedov solution has generated.

**Review by sekelle — CHANGES_REQUESTED** [review]
Hi @jaescartin1, thanks for your contribution.
The analytical solution itself and the changes to the SedovDataGenerator are fine.

One problem that remains is that the code for the analytical solution is introduced into the main compilation unit of sedov.cpp.
We have to avoid this coupling, because what's now just the sedov test case in sedov.cpp will evolve into a more general simulation driver that will be able to handle any hydro-only type of simulation, including the turbulence tests on LUMI.
We don't actually want to repeat the whole time-stepping loop for every new set of initial conditions. Therefore we can't have
code for the analytical solution introduced in sedov.cpp

The sedov analytical solution can and should be a stand-alone application: You first execute a run with the sedov application, writing snapshots at the times where you want to compare with the analytical solutions.
Next, sedov-analytical can be executed. It includes the same sedov_data_generator.hpp file as sedov.cpp to get the simulation parameters and it reads the required simulation data (x,y,z,vx,vy,vz,time) from the output that the sedov solution has generated.

**Comment by sekelle** [comment]
Note, I would set up `sedov-analytical` such that it compares only a single time/snapshot and control comparing multiple times with a python script.

File: src/sedov/SedovDataGenerator.hpp
Comment (sekelle) [inline src/sedov/SedovDataGenerator.hpp]: How is this not generating a compiler error in the CI? This should not compile because `std::pow` is not constexpr,
and in fact does not compile for me.

Comment (jaescartin1) [reply src/sedov/SedovDataGenerator.hpp]: I couldn't see this problem with make or cmake in my Ubuntu.
Do you use some additional flags for the compilation?

Maybe is a different compiler version. This is mine:
$ mpic++ --version
g++ (Ubuntu 11.2.0-7ubuntu2) 11.2.0

Comment (jaescartin1) [reply src/sedov/SedovDataGenerator.hpp]: As far as I can see on my system, the pow(x,y) function is taken from the include <math.h> at '/usr/x86_64-linux-gnu/include/bits/mathcalls.h'

$ cat /usr/x86_64-linux-gnu/include/bits/mathcalls.h
...
/* Return X to the Y power.  */
__MATHCALL_VEC (pow,, (_Mdouble_ __x, _Mdouble_ __y));
...

Comment (sekelle) [reply src/sedov/SedovDataGenerator.hpp]: It works with GCC and fails with clang. According to the C++17 standard, std::pow is not constexpr, so GCC is wrong and clang is right. This should be:
`static inline const double energ0 = ...`

Comment (sekelle) [reply src/sedov/SedovDataGenerator.hpp]: @jgphpc are we only testing with GCC?

Comment (jgphpc) [reply src/sedov/SedovDataGenerator.hpp]: yes but I can add clang

**Comment by jgphpc** [comment]
cscs-ci run

**Comment by jaescartin1** [comment]
The required changes are in-place. Please, check it again.

**Comment by jgphpc** [comment]
cscs-ci run

File: src/analyticalSolutions/compare_solutions.py
Comment (sekelle) [inline src/analyticalSolutions/compare_solutions.py]: perhaps `sedov_pressure` is enough (note double ss) ?

**Comment by sekelle** [comment]
Some observations:

- An SVG file with 125000 (and more) particles is computationally very demanding to render, we should perhaps make .png the default
- File naming convention: the most widely used naming scheme is `lower_case_with_underscore`, e.g. `sedov_solution.cpp`, I already applied this convention everywhere in `domain/`, but not yet in `include/`
- I invoked the python script with `./compare_solutions.py sedov --binary_file ./sedovSolution --nparts 125000 --snapshot_file ./dump_sedov200.txt --constants_file ./constants.txt`, i.e. I forgot to also supply `--iteration 200`. In this case the script assumed iteration 0 as the default, so I got a wrong plot. Perhaps we should make `iteration` a mandatory argument if `constants_file` is given

**Comment by sekelle** [comment]
When building with cmake, the `compare_solutions.py` script is not copied to the installation directory, so I had to copy that by hand. Perhaps @jgphpc could advise how to modify CMakeLists.txt to also install the python script?

File: src/analyticalSolutions/compare_solutions.py
Comment (jgphpc) [inline src/analyticalSolutions/compare_solutions.py]: comparison

File: src/analyticalSolutions/sedovSolution/main.cpp
Comment (jgphpc) [inline src/analyticalSolutions/sedovSolution/main.cpp]: You may want to add a reference:
https://cococubed.com/papers/kamm_2000.pdf

Comment (jaescartin1) [reply src/analyticalSolutions/sedovSolution/main.cpp]: Hi, yes sure, but this is already added but not in the 'main.ccp' where only is called. The reference is in the solution header itself 'SedovSolution.hpp'

 *        This routine is a C++ conversion of one Fortran code based in these two papers:
 *        - "evaluation of the sedov-von neumann-taylor blast wave solution", Jim Kamm, la-ur-00-6055
 *        - "the sedov self-similiar point blast solutions in nonuniform media", David Book, shock waves, 4, 1, 1994

Comment (jaescartin1) [reply src/analyticalSolutions/sedovSolution/main.cpp]: The Frank Timmes code (in the source that you pointed out) is a evolution, wrote in 2019, of the base code that I used.
https://cococubed.com/research_pages/sedov.shtml

Comment (jaescartin1) [reply src/analyticalSolutions/sedovSolution/main.cpp]: I ported and removed the bad practices (as goto jumps!) but maybe need a more deep refactoring to be more nice and readable.

Comment (jaescartin1) [reply src/analyticalSolutions/sedovSolution/main.cpp]: Checking the results in 1D, 2D and 3D, both codes generate the same outputs but with a small diferences done by the numerical noise introduced for the conversion from Fortran to C++

Comment (jaescartin1) [reply src/analyticalSolutions/sedovSolution/main.cpp:28]: Checking the results in 1D, 2D and 3D, both codes generate the same outputs but with a small differences done by the numerical noise introduced for the conversion from Fortran to C++

**Comment by jgphpc** [comment]
> the `compare_solutions.py` script is not copied to the installation directory

see https://github.com/unibas-dmi-hpc/SPH-EXA/pull/140/commits/c58d9ecadf51dead99474f743fe2cd132f46a835

**Comment by jgphpc** [comment]
cscs-ci run

File: src/analyticalSolutions/compare_solutions.py
Comment (sekelle) [reply src/analyticalSolutions/compare_solutions.py]: also here, perhaps we can just use `sedov_density_` for brevity.

**Comment by jgphpc** [comment]
cscs-ci run

File: src/analyticalSolutions/sedovSolution/CMakeLists.txt
Comment (jgphpc) [inline src/analyticalSolutions/sedovSolution/CMakeLists.txt]: There is a failure when building `sedovSolution` with clang++:
```
ld.lld: error: undefined symbol: std::filesystem::__cxx11::path::_M_split_cmpts()
>>> referenced by main.cpp
```

This fixes the issue:
```suggestion
target_link_libraries(sedovSolution stdc++fs)
install(TARGETS sedovSolution RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
```

**Comment by jaescartin1** [comment]
Hi @sebkelle1 ,
     I set the iteration value to -1. Like first is check if time>0.0 (default value time=0.0),  if you don't specify time and you give the 'constant.txt' file instead but without iteration you will receive an error.
It make sense to check the iteration=0, if you want to compare the model at time=0. with the analytical solutión
... but I think there is an issue in the 'constants.txt' .  
iter=0 is the initial snapshot, without any movement but the constants file said that the time is t=dt, instead t=0.
Maybe you added t=t+dt before to write the line in the constants file.

**Comment by jaescartin1** [comment]
$ cat constants.txt 
0 1.1e-06 1.1e-06 1 1.00739e-08 1 0 12702448 
1 2.31e-06 1.21e-06 1 3.10683e-08 1 0 12697729 
...

**Comment by jaescartin1** [comment]
All the changes proposed in your comments are in-place.
Additionally I added the L1 error control for the common variables: rho, u, p, vel, cs
Please, review again the changes

**Comment by jaescartin1** [comment]
Here the execution commands for the new version. 
All the output files will be generated in ./bin/
The pictures will be in .png format.
The call also check the rho L1 error

$ git clone --branch "analytical_sedov" https://github.com/jaescartin1/SPH-EXA.git
$ cd SPH-EXA && make -j mpi+omp && ./bin/mpi+omp.app -n 50 -s 200 -w 200 --outDir ./bin/
$ python src/analytical_solutions/compare_solutions.py sedov --binary_file bin/sedov_solution --constants_file ./bin/constants.txt --iteration 200 --nparts 125000 --snapshot_file ./bin/dump_sedov200.txt --out_dir bin/ --error_rho --delta_rho 1.e-0

**Comment by jgphpc** [comment]
> I cannot see the python file in the build/ folder. Am I doing something wrong?

no, it will copy it only when running `make install`.
That said, I need to fix an issue in cmake (I'll push).

Btw, `-DBUILD_ANALYTICAL=ON` is on by default.

**Comment by jgphpc** [comment]
cscs-ci run

**Comment by sekelle** [comment]
cscs-ci run

**Comment by sekelle** [comment]
This is ready for merging from my side. As soon as the CI is available again we can merge.

File: src/analytical_solutions/sedov_solution/sedov_solution.hpp
Comment (sekelle) [inline src/analytical_solutions/sedov_solution/sedov_solution.hpp]: We don't need these include guards. We have "pragma once" just below

File: src/analytical_solutions/sedov_solution/particle.hpp
Comment (sekelle) [inline src/analytical_solutions/sedov_solution/particle.hpp]: not needed

File: src/analytical_solutions/sedov_solution/io.hpp
Comment (sekelle) [inline src/analytical_solutions/sedov_solution/io.hpp]: not needed

File: src/analytical_solutions/sedov_solution/main.cpp
Comment (jgphpc) [inline src/analytical_solutions/sedov_solution/main.cpp:33]:
```suggestion
#include <sstream>
#include <iostream>
```

File: src/analytical_solutions/sedov_solution/io.cpp
Comment (jgphpc) [inline src/analytical_solutions/sedov_solution/io.cpp:26]:
```suggestion
#include <sstream>
#include <iostream>
```

File: src/analytical_solutions/sedov_solution/CMakeLists.txt
Comment (jgphpc) [inline src/analytical_solutions/sedov_solution/CMakeLists.txt]:
```suggestion
if(CMAKE_OSX_SYSROOT)
    target_link_libraries(sedov_solution stdc++)
else()
     target_link_libraries(sedov_solution stdc++fs)
endif()
```

**Comment by jaescartin1** [comment]
Ruben is checking the Error L1 in u and cs. Let's wait for the merge till Ruben give the ok

**Comment by jgphpc** [comment]
⚠️: I pushed compare_solutions.py (reformatting only, no change to the code)

**Comment by jaescartin1** [comment]
Perfect, I put my last changes in your reformatted code

**Review by sekelle — APPROVED** [review]

**Review by jgphpc — APPROVED** [review]

**Pull request merged**
